packages/@uppy/companion/src/server/controllers/send-token.js

```diff
@@ -1,12 +1,27 @@
-/**
- *
- * sends auth token to uppy client
- */
 const { URL } = require('url')
 const tokenService = require('../helpers/jwt')
 const { hasMatch, sanitizeHtml } = require('../helpers/utils')
 const oAuthState = require('../helpers/oauth-state')
-const versionCmp = require('../helpers/version')
+
+/**
+ *
+ * @param {string} token uppy auth token
+ * @param {string} origin url string
+ */
+const htmlContent = (token, origin) => {
+  return `
+    <!DOCTYPE html>
+    <html>
+    <head>
+        <meta charset="utf-8" />
+        <script>
+          window.opener.postMessage(JSON.stringify({token: "${token}"}), "${sanitizeHtml(origin)}")
```

Comment:
> `JSON.stringify` I believe is very unsafe

Actually using `JSON.stringify` would be much safer, see my suggestion below.

Reply:
Is it really safer?

If token is something like `JSON.stringify('</script><script>console.log(JSON.stringify(localStorage))</script>')`, then it would allow executing custom dangerous code, no? It results in:
```
"\"</script><script>console.log(JSON.stringify(localStorage))</script>\""
```

Reply:
Good catch, if the string contains a close script tag, that's an issue. Solution to this in https://pragmaticwebsecurity.com/articles/spasecurity/json-stringify-xss.html (`JSON.stringify(token).replaceAll('<', '\\u003c')`), I've updated my suggestion below.

Reply:
we could pull the token through `sanitizeHtml` too. the JSON.stringify in this snippet happens on the client side

Reply:
I'm not sure if sanitizeHtml is safe to use inside `<script>` tags. There are very strict and somewhat complex rules for what can be safely put inside a script ta. See https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#rule-3-javascript-encode-before-inserting-untrusted-data-into-javascript-data-values

https://github.com/transloadit/uppy/blob/f4e99fa7bf8ac085d4215215aa1d2b3342ef9fef/packages/%40uppy/companion/src/server/helpers/utils.js#L40

Reply:
I think the only easy safe things to do is to either
1. escape the token/origin as html and put it inside a hidden DIV as JSON.stringified, then use javascript in the document to parse the div content before using it.
2. if the token/origin only consists of a certain type of characters (e.g. alphanumeric), then do a `token.replace(/[^0-9a-z]/ig, '')` before inserting it to the document. then it's guaranteed to not contain any unsafe characters.